File: drones/drone.py
MAX_X = 1366
MAX_Y = 768
MAX_Z = 100

class Drone:
    """This is a drone class"""
    
    all_drones = []

    def __init__(self, name: str, x: int, y: int, z: int):
        assert 0 <= x < MAX_X, f"x coordinate must be in range (0, {MAX_X})"
        assert 0 <= y < MAX_Y, f"y coordinate must be in range (0, {MAX_Y})"
        assert 0 <= z < MAX_Z, f"z coordinate must be in range (0, {MAX_Z})"
        
        self.__name = name
        self.__x = x
        self.__y = y
        self.__z = z
        Drone.all_drones.append(self)

    @property
    def coordinates(self):
        """Get cordinates"""
        return self.__x, self.__y, self.__z

    @coordinates.setter
    def coordinates(self, new_coordinates):
        x, y, z = new_coordinates
        if not (0 <= x < MAX_X):
            raise ValueError(f"x coordinate must be in range (0, {MAX_X})")
        if not (0 <= y < MAX_Y):
            raise ValueError(f"y coordinate must be in range (0, {MAX_Y})")
        if not (0 <= z < MAX_Z):
            raise ValueError(f"z coordinate must be in range (0, {MAX_Z})")
        self.__x, self.__y, self.__z = x, y, z

    def __repr__(self):
        return f"Drone(name={self.__name}, x={self.__x}, y={self.__y}, z={self.__z})"
    
    def __str__(self):
        return self.__name

File: drones/drone.cpp
#include <string>
#include <iostream>
#include <cassert>
#include <tuple>
#include <vector>

const int MAX_X = 1366;
const int MAX_Y = 768;
const int MAX_Z = 100;

class AbstractClass{
    virtual void collision() = 0;
};

class Drone: AbstractClass {
private:
    std::string name;
    int x_cor, y_cor, z_cor;

public:
    //Public variables
    std::vector<std::string> *drone_location;
    // Constructor
    Drone(std::string aName, int aX, int aY, int aZ) {
        assert(aX > 0 && aX < MAX_X && "x must be positive and in range");
        assert(aY > 0 && aY < MAX_Y && "y must be positive and in range");
        assert(aZ > 0 && aZ < MAX_Z && "z must be positive and in range");
        name = aName;
        x_cor = aX;
        y_cor = aY;
        z_cor = aZ;
        
    }

    //Collision contract implementation
    void collision(){

    }

    //Get coordinates and return tuples
    std::tuple<int, int, int> getCordinates() {
        return std::make_tuple(x_cor, y_cor, z_cor);
    }

    std::string getName() {
        return name;
    }

    void setCordinates(int aX, int aY, int aZ) {
        x_cor = aX;
        y_cor = aY;
        z_cor = aZ;
        std::cout << "Coordinates updated successfully" << std::endl;
    }

    
};

int main() {
    Drone drone("drone1", 239, 33, 32);
    std::cout << "Drone Name: " << drone.getName() << std::endl;

    // Print coordinates
    auto [x, y, z] = drone.getCordinates();
    std::cout << "Coordinates: (" << x << ", " << y << ", " << z << ")" << std::endl;

    return 0;
}
